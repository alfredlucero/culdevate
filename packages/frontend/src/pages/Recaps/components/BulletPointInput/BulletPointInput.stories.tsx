import React from "react";
import { storiesOf } from "@storybook/react";
import BulletPointInput from "./index";

const onClickDelete = () => {
  console.log("Clicked delete bullet point!");
};
const onChange = () => {
  console.log("Changed bullet point input!");
};
const onBlur = () => {
  console.log("Blurred bullet point input!");
};

storiesOf("RecapsPage/BulletPointInput", module)
  .add("Default Empty", () => (
    <BulletPointInput
      onChange={onChange}
      onBlur={onBlur}
      onClickDelete={onClickDelete}
      id="bulletpoint1"
      value=""
      valid={true}
      errorInfo="Error info"
    />
  ))
  .add("Valid Input", () => (
    <BulletPointInput
      onChange={onChange}
      onBlur={onBlur}
      onClickDelete={onClickDelete}
      id="bulletpoint1"
      value="valid bullet point"
      valid={true}
      errorInfo="Error info"
    />
  ))
  .add("Invalid Input", () => (
    <BulletPointInput
      onChange={onChange}
      onBlur={onBlur}
      onClickDelete={onClickDelete}
      id="bulletpoint1"
      value="invalid bullet point"
      valid={false}
      errorInfo="Bullet point is invalid"
    />
  ))
  .add("Multiple Inputs", () => (
    <div>
      <BulletPointInput
        onChange={onChange}
        onBlur={onBlur}
        onClickDelete={onClickDelete}
        id="bulletpoint1"
        value=""
        valid={true}
        errorInfo=""
      />
      <BulletPointInput
        onChange={onChange}
        onBlur={onBlur}
        onClickDelete={onClickDelete}
        id="bulletpoint1"
        value="valid bullet point"
        valid={true}
        errorInfo=""
      />
      <BulletPointInput
        onChange={onChange}
        onBlur={onBlur}
        onClickDelete={onClickDelete}
        id="bulletpoint1"
        value="invalid bullet point"
        valid={false}
        errorInfo="Bullet point is invalid"
      />
    </div>
  ));
