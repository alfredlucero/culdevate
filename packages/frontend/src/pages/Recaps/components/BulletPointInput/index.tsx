import React from "react";
import cn from "classnames";
import TextInput, { TextInputProps } from "../../../../components/TextInput";
import Icon from "../../../../components/Icon";
import { CommonProps } from "../../../../components/commonProps";

interface BulletPointInputProps extends CommonProps {
  onClickDelete: (e: React.MouseEvent) => void;
  onChange: TextInputProps["onChange"];
  onBlur: TextInputProps["onBlur"];
  value: TextInputProps["value"];
  id: TextInputProps["id"];
  valid: TextInputProps["valid"];
  errorInfo: TextInputProps["errorInfo"];
}

const BulletPointInput: React.FC<BulletPointInputProps> = ({
  onClickDelete,
  onChange,
  onBlur,
  value,
  id,
  valid = true,
  errorInfo = "",
  testId = "",
  className = "",
  ...passThroughProps
}) => {
  return (
    <div
      {...(testId !== "" ? { "data-testid": testId } : {})}
      className={cn(
        "flex",
        "items-start",
        "px-4",
        "py-6",
        "rounded",
        "hover:shadow-md",
        "hover:bg-blue-100",
        "cursor-pointer",
        className,
      )}
      {...passThroughProps}
    >
      <TextInput
        onChange={onChange}
        onBlur={onBlur}
        value={value}
        id={id}
        valid={valid}
        errorInfo={errorInfo}
        required={true}
        type="text"
        placeholder="Provide your highlights and results..."
        testId={testId ? `${testId}Input` : ""}
        className={cn("mr-4", "flex-1", "cursor-none")}
      />
      <Icon
        variant="trash"
        size="medium"
        onClick={onClickDelete}
        className={cn("text-red-700", "hover:text-red-900", "cursor-pointer")}
        testId={testId ? `${testId}DeleteIcon` : ""}
      />
    </div>
  );
};

export default BulletPointInput;
