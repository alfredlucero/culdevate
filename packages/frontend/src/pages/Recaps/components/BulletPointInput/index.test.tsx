import React from "react";
import { render, fireEvent } from "@testing-library/react";
import BulletPointInput from "./index";

describe("<BulletPointInput />", () => {
  const bulletPointInputTestId = "bulletPointInputTestId";

  test("should render without error", () => {
    const { container } = render(
      <BulletPointInput
        onChange={() => {}}
        onBlur={() => {}}
        onClickDelete={() => {}}
        id="bulletpointid"
        value="bullet point description"
        valid={true}
        errorInfo=""
        testId={bulletPointInputTestId}
        className="extraBulletPointClass"
      />,
    );

    expect(container).toMatchSnapshot();
  });

  test("should display inline error given invalid input and errors", () => {
    const errorInfo = "Bullet point is invalid";
    const { getByTestId } = render(
      <BulletPointInput
        onChange={() => {}}
        onBlur={() => {}}
        onClickDelete={() => {}}
        id="bulletpointid"
        value="invalid bullet point"
        valid={false}
        errorInfo={errorInfo}
        testId={bulletPointInputTestId}
      />,
    );

    expect(getByTestId(bulletPointInputTestId).textContent).toContain(errorInfo);
  });

  test("should not display inline error given valid input", () => {
    const errorInfo = "Bullet point is invalid";
    const { getByTestId } = render(
      <BulletPointInput
        onChange={() => {}}
        onBlur={() => {}}
        onClickDelete={() => {}}
        id="bulletpointid"
        value="valid bullet point"
        valid={true}
        errorInfo="Error info"
        testId={bulletPointInputTestId}
      />,
    );

    expect(getByTestId(bulletPointInputTestId).textContent).not.toContain(errorInfo);
  });

  test("should call onClickDelete when clicking the input's delete icon", () => {
    const onClickDeleteMock = jest.fn();
    const { getByTestId } = render(
      <BulletPointInput
        onChange={() => {}}
        onBlur={() => {}}
        onClickDelete={onClickDeleteMock}
        id="bulletpointid"
        value="valid bullet point"
        valid={true}
        errorInfo="Error info"
        testId={bulletPointInputTestId}
      />,
    );

    fireEvent.click(getByTestId(`${bulletPointInputTestId}DeleteIcon`));

    expect(onClickDeleteMock).toHaveBeenCalled();
  });
});
