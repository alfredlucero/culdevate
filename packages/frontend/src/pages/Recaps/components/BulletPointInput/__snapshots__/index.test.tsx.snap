// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BulletPointInput /> should render without error 1`] = `
<div>
  <div
    class="flex items-start px-4 py-6 rounded hover:shadow-md hover:bg-blue-100 cursor-pointer extraBulletPointClass"
    data-testid="bulletPointInputTestId"
  >
    <div
      class="mr-4 flex-1 cursor-none"
    >
      
      <input
        aria-required="true"
        class="shadow appearance-none border w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline rounded"
        data-testid="bulletPointInputTestIdInput"
        id="bulletpointid"
        name="bulletpointid"
        placeholder="Provide your highlights and results..."
        type="text"
        value="bullet point description"
      />
    </div>
    <div
      class="font-sans text-gray-800 text-2xl text-red-700 hover:text-red-900 cursor-pointer"
      data-testid="bulletPointInputTestIdDeleteIcon"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-trash-alt fa-w-14 "
        data-icon="trash-alt"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M32 464a48 48 0 0 0 48 48h288a48 48 0 0 0 48-48V128H32zm272-256a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zM432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z"
          fill="currentColor"
        />
      </svg>
    </div>
  </div>
</div>
`;
